### PR TITLE
feat(templates): integrate Managed Kafka notification templates

### PR DESCRIPTION
**Pull request by machi1990**
This is the first step of integrating Managed Kafka templates. We'll be working on adding some
content on each of the added templates but let's keep them empty for now to be able to perform all
the required Java bindings

/cc @gwenneg @akoserwal @theute

### CODE REVIEW

**Comment by machi1990** [comment]
It is my first time opening a PR in this repo, I might have missed the some required steps so looking for guidance. @gwenneg

File: backend/src/main/java/com/redhat/cloud/notifications/templates/EmailTemplateFactory.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/templates/EmailTemplateFactory.java:28]: I don't know why we used `equalsIgnoreCase` for other templates but could you please do the same here? I'll have a look later to see if this `equalsIgnoreCase` can be replaced with `equals`.

File: backend/src/main/resources/templates/Rhosak/actionRequredBody.html
Comment (gwenneg) [inline backend/src/main/resources/templates/Rhosak/actionRequredBody.html:2]: There's a typo in this file name 😃 `actionRequredBody` -> `actionRequiredBody`

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java:28]: @josejulio Returning `null` won't cause any issue in the calling code, but I can see other templates throwing an exception here. What's your preference?

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java:44]: @josejulio Same question here.

**Review by gwenneg — COMMENTED** [review]
So far it looks good to me!

File: backend/src/main/resources/templates/Rhosak/actionRequredBody.html
Comment (machi1990) [reply backend/src/main/resources/templates/Rhosak/actionRequredBody.html:2]: Good catch, I knew I was bound to make a mistake on the name.

File: backend/src/main/java/com/redhat/cloud/notifications/templates/EmailTemplateFactory.java
Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/EmailTemplateFactory.java:28]: @theute confirmed that the name is `rhosak` that's why I used equal but no harm in ignoring case.

File: backend/src/main/resources/templates/Rhosak/actionRequredBody.html
Comment (machi1990) [reply backend/src/main/resources/templates/Rhosak/actionRequredBody.html:2]: Thanks, fixed.

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java:28]: Good point, technically we should not arrive here. I think I'll go with the exception for consistency.

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/aggregators/RhosakEmailAggregator.java
Comment (machi1990) [inline backend/src/main/java/com/redhat/cloud/notifications/processors/email/aggregators/RhosakEmailAggregator.java:8]: @gwenneg @josejulio I see that CI is complaining because of a constraint violation of payload as it cannot be `null`. I suspect it is coming from here. For now we do not have a template defined yet but that's in coming work so for now we are only looking to put the infrastructural code in. Is there a way to trick it so that the payload is not null?

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java]: I've changed it to throw an exception instead for consistency.

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java]: Fixed.

**Comment by machi1990** [comment]
@gwenneg thanks for the review. I left a question on https://github.com/RedHatInsights/notifications-backend/pull/576#discussion_r705183607 regarding one test failure I saw. I suspect it is coming from the empty aggregator that I've added. Looking for expert input on this.

File: backend/src/main/java/com/redhat/cloud/notifications/templates/EmailTemplateFactory.java
Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/EmailTemplateFactory.java]: Changed to ignoring cases for now.

File: backend/src/main/resources/templates/Rhosak/actionRequiredBody.html
Comment (gwenneg) [inline backend/src/main/resources/templates/Rhosak/actionRequiredBody.html:1]: CI is complaining because all `{#include ...}` tags must be closed with `{/include}`.

Comment (machi1990) [reply backend/src/main/resources/templates/Rhosak/actionRequiredBody.html:1]: I am looking at this now. I want to put some skeleton in the initial templates. 

My question was on the following exception
```
Couldn't persist aggregation!: javax.validation.ConstraintViolationException: Validation failed for classes [com.redhat.cloud.notifications.models.EmailAggregation] during persist time for groups [javax.validation.groups.Default, ]
List of constraint violations:[
	ConstraintViolationImpl{interpolatedMessage='must not be null', propertyPath=accountId, rootBeanClass=class com.redhat.cloud.notifications.models.EmailAggregation, messageTemplate='{javax.validation.constraints.NotNull.message}'}
]
```

or it is not something I've to worry about?

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/aggregators/RhosakEmailAggregator.java
Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/aggregators/RhosakEmailAggregator.java:8]: I was seeing the following exception in the logs
```
Couldn't persist aggregation!: javax.validation.ConstraintViolationException: Validation failed for classes [com.redhat.cloud.notifications.models.EmailAggregation] during persist time for groups [javax.validation.groups.Default, ]
List of constraint violations:[
	ConstraintViolationImpl{interpolatedMessage='must not be null', propertyPath=accountId, rootBeanClass=class com.redhat.cloud.notifications.models.EmailAggregation, messageTemplate='{javax.validation.constraints.NotNull.message}'}
]
```

Wondering if this is something known or it is a result of this change.

File: backend/src/main/resources/templates/Rhosak/actionRequiredBody.html
Comment (gwenneg) [reply backend/src/main/resources/templates/Rhosak/actionRequiredBody.html]: Don't worry about this warn, it's not related to your PR. It comes from here:
https://github.com/RedHatInsights/notifications-backend/blob/2ed05771a0a37a74518c1896c5e7c0cfd9f5af2f/backend/src/test/java/com/redhat/cloud/notifications/db/EmailAggregationResourcesTest.java#L81-L92

File: backend/src/main/java/com/redhat/cloud/notifications/processors/email/aggregators/RhosakEmailAggregator.java
Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/processors/email/aggregators/RhosakEmailAggregator.java:8]: Answered here https://github.com/RedHatInsights/notifications-backend/pull/576#discussion_r705210051

File: backend/src/main/resources/templates/Rhosak/actionRequiredBody.html
Comment (machi1990) [reply backend/src/main/resources/templates/Rhosak/actionRequiredBody.html]: Perfect, thanks. I guess we are ready to promote the PR for full review. Thank you again for the initial review @gwenneg

File: backend/src/main/java/com/redhat/cloud/notifications/templates/EmailTemplateFactory.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/templates/EmailTemplateFactory.java:35]:
```suggestion
        } else if (bundle.equalsIgnoreCase(APPLICATION_SERVICES)) {
```

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (gwenneg) [inline backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java]: Are you OK with a dash separator in the application name while the event types names are separated with underscores? This is not what you initially requested so I want to make sure it's fine 😃

File: backend/src/main/resources/templates/Rhosak/insightsEmailBody.html
Comment (gwenneg) [inline backend/src/main/resources/templates/Rhosak/insightsEmailBody.html:487]:
```suggestion
                                        This email was sent by Red Hat Insights |<a href="https://console.redhat.com/user-preferences/notifications/application-services" target="_blank">&nbsp;Manage notification preferences</a>
```

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (gwenneg) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java]: Oh you addressed that question that while I was reviewing 😄

File: backend/src/main/resources/templates/Rhosak/insightsEmailBody.html
Comment (machi1990) [reply backend/src/main/resources/templates/Rhosak/insightsEmailBody.html:487]: Good catch.

File: backend/src/main/java/com/redhat/cloud/notifications/templates/EmailTemplateFactory.java
Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/EmailTemplateFactory.java]: Thanks, done.

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java]: Yes, I saw @theute comment in the JIRA and applied the changes. Good spot as usual by the way.

File: backend/src/main/resources/templates/Rhosak/insightsEmailBody.html
Comment (machi1990) [reply backend/src/main/resources/templates/Rhosak/insightsEmailBody.html]: done

**Review by gwenneg — APPROVED** [review]
LGTM

Let's wait for @josejulio's approval before the merge.

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (josejulio) [inline backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java:27]: `eventType` is `null` on daily aggregation. As it aggregates over all the `eventTypes` found.
Since `SCHEDULED_UPGRADE` is only meant to be used for daily aggregations I think it should be changed with something like:

```suggestion
        if (type == EmailSubscriptionType.INSTANT) {
            if (eventType.equals(DISRUPTION)) {
                return Templates.serviceDisruptionTitle();
            } else if (eventType.equals(INSTANCE_CREATED)) {
                return Templates.instanceCreatedTitle();
            } else if (eventType.equals(INSTANCE_DELETED)) {
                return Templates.instanceDeletedTitle();
            } else if (eventType.equals(ACTION_REQUIRED)) {
                return Templates.actionRequiredTitle();
            }
        } else if (type == EmailSubscriptionType.DAILY) {
            return Templates.dailyTitle();
        }
```

The `scheduledUpgradeTitle` template should probably be removed (or renamed) and a template for the daily aggregation should be created/used instead

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (josejulio) [inline backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java:47]: Similar as above

```suggestion
        if (type == EmailSubscriptionType.INSTANT) {
            if (eventType.equals(DISRUPTION)) {
                return Templates.serviceDisruptionBody();
            } else if (eventType.equals(INSTANCE_CREATED)) {
                return Templates.instanceCreatedBody();
            } else if (eventType.equals(INSTANCE_DELETED)) {
                return Templates.instanceDeletedBody();
            } else if (eventType.equals(ACTION_REQUIRED)) {
                return Templates.actionRequiredBody();
            }
        } else if (type == EmailSubscriptionType.DAILY) {
            return Templates.dailyBody();
        }
```

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (josejulio) [inline backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java:58]: IIUC, `SCHEDULED_UPGRADE` won't have an instant email, but will be part of the daily email?

**Review by josejulio — CHANGES_REQUESTED** [review]

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java:27]: Why the change of template name to be prefixed with `dailyXX`?

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java:47]: Thanks for the review @josejulio the suggestion looks good to me other than the rename of `scheduledUpgradeBody` to `dailyBody` which to me is more generic and I'll rather it stays specific and more explicit to what the template is about rather than email frequency / type.

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (josejulio) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java:27]: Because, as we currently have, all the event types are aggregated into a single payload and sent as one daily email.
In this case (daily) we don't have different emails for each event.

The name can be anything you want, I just choose `dailyTitle` as an example.

Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java:27]: @akoserwal is this what we want? i.e aggregate all daily emails and send them over one email?

Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java:27]: > Because, as we currently have, all the event types are aggregated into a single payload and sent as one daily email.
> In this case (daily) we don't have different emails for each event.
> 
> The name can be anything you want, I just choose `dailyTitle` as an example.

Thanks @josejulio , let's get a confirmation about the behaviour we want and I'll do the necessary updates. Based on your description of aggregation I am not sure if this is what we want.

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (josejulio) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java:47]: Any name you pick is fine, I just want to clarify that in our current processing a single daily email is sent with all the aggregated event_types (You can choose what to include or not)

Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java:47]: Ah perfect thank you for pointing it out. I think what you are proposing as a name makes sense. I want to confirm first if this is what we expect i.e daily email of all event types rather than a daily email for each event type. 

/cc @akoserwal @davidffrench

**Comment by gwenneg** [comment]
I knew my LGTM wasn't good enough, I'm still learning from the templates master (aka @josejulio) 😄

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (davidffrench) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java]: Both are very valid approaches to be honest, it comes down to personal preference. 

With the current architecture of the notification service where notifications are sent to all users in an organisation, this makes sense. However, in the future, once https://issues.redhat.com/browse/NOTIF-205 and https://issues.redhat.com/browse/NOTIF-206 have been completed, this will be more nuanced. If the logic for sending a single daily notification in this case will only include the notifications sent to the user during the day, this approach makes perfect sense.

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java]: I've made the change with the daily template.

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java:66]: I've confirmed it and indeed this is what we want.

File: backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java
Comment (machi1990) [reply backend/src/main/java/com/redhat/cloud/notifications/templates/Rhosak.java]: Perfect, thanks David. 

I have went ahead with the proposed change.

@josejulio do you want to have another look?

**Comment by machi1990** [comment]
> I knew my LGTM wasn't good enough, I'm still learning from the templates master (aka @josejulio)

I am glad I got @josejulio review too.

**Review by josejulio — APPROVED** [review]
lgtm

**Comment by machi1990** [comment]
Thanks for the approval @josejulio @gwenneg, is there anything needed to be done for the work to be merged?

**Comment by josejulio** [comment]
> is there anything needed to be done for the work to be merged?

No, lets merge it now :)

**Comment by machi1990** [comment]
Awesome thanks @josejulio it should be available in `ci.console.redhat` env for testing right?

**Comment by josejulio** [comment]
> Awesome thanks @josejulio it should be available in `ci.console.redhat` env for testing right?

sorry, missed the comment. 
You are correct, it should be already on ci. (takes some minutes)

**Comment by machi1990** [comment]
> > Awesome thanks @josejulio it should be available in `ci.console.redhat` env for testing right?
> 
> sorry, missed the comment.
> You are correct, it should be already on ci. (takes some minutes)

Thanks

**Pull request merged**
